connection.go

```diff
@@ -1233,6 +1359,224 @@ func (conn *Connection) NewStream() (*Stream, error) {
+			<-state.unready
+			st = nil
+		}
+	}
+
+	return st, nil
+}
+
+// NewWatcher creates a new Watcher object for the connection.
+//
+// You need to require WatchersFeature to use watchers, see examples for the
+// function.
+//
+// After watcher creation, the watcher callback is invoked for the first time.
+// In this case, the callback is triggered whether or not the key has already
+// been broadcast. All subsequent invocations are triggered with
+// box.broadcast() called on the remote host. If a watcher is subscribed for a
+// key that has not been broadcast yet, the callback is triggered only once,
+// after the registration of the watcher.
+//
+// The watcher callbacks are always invoked in a separate goroutine. A watcher
+// callback is never executed in parallel with itself, but they can be executed
+// in parallel to other watchers.
+//
+// If the key is updated while the watcher callback is running, the callback
+// will be invoked again with the latest value as soon as it returns.
+//
+// Watchers survive reconnection. All registered watchers are automatically
+// resubscribed when the connection is reestablished.
+//
+// Keep in mind that garbage collection of a watcher handle doesn’t lead to the
+// watcher’s destruction. In this case, the watcher remains registered. You
+// need to call Unregister() directly.
+//
+// Unregister() guarantees that there will be no the watcher's callback calls
+// after it, but Unregister() call from the callback leads to a deadlock.
+//
+// See:
+// https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_events/#box-watchers
+//
+// Since 1.10.0
+func (conn *Connection) NewWatcher(key string, callback WatchCallback) (Watcher, error) {
+	// We need to check the feature because the IPROTO_WATCH request is
+	// asynchronous. We do not expect any response from a Tarantool instance
+	// That's why we can't just check the Tarantool response for an unsupported
+	// request error.
+	watchersRequired := false
+	for _, feature := range conn.opts.RequiredProtocolInfo.Features {
+		if feature == WatchersFeature {
+			watchersRequired = true
+			break
+		}
+	}
+
+	if !watchersRequired {
+		err := fmt.Errorf("the feature %s must be required by connection "+
+			"options to create a watcher", WatchersFeature)
+		return nil, err
+	}
+
+	st, err := subscribeWatchChannel(conn, key)
+	if err != nil {
+		return nil, err
+	}
+
+	// Start the watcher goroutine.
+	done := make(chan struct{})
+	finished := make(chan struct{})
+
+	go func() {
+		var version uint64 = initWatchEventVersion
+		for {
+			state := <-st
+			if state.changed == nil {
+				state.changed = make(chan struct{})
+			}
+			st <- state
+
+			if state.version != version {
+				callback(WatchEvent{
+					Conn:  conn,
+					Key:   key,
+					Value: state.value,
+				})
+				version = state.version
+
+				// Do we need to acknowledge the notification?
+				state = <-st
+				ack := !state.ack && version == state.version
```

Comment:
`ack := !state.ack` is counter-intuitive

Reply:
Renamed: `ack` -> `sendAck`